### PR TITLE
new flag --listen-apiserver-port  for docker/podman: allow user to specify host port for api-server  

### LINKED ISSUE #11041: apiserver port is exposed to a random port
Hi,
Currently, the exposed port will be changed once minikube is restarted.

If apiserver is exposed on a fixed port, it would be convenient to access this cluster remotely, that's to say, there is no need to modify the cluster information in kubeconfig file once minikube is restarted.

**Steps to reproduce the issue:** 

1. $ minikube start --driver=docker --listen-address='0.0.0.0' --apiserver-ips=172.28.24.96 
2. $ docker ps
    Get the exposed apiserver port. It's "0.0.0.0:32823->8443/tcp".
3. $ minikube stop
4. $ minikube start
4. $ docker ps
    Get the exposed apiserver port. It's "0.0.0.0:32828->8443/tcp".

**Full output of `minikube start` command used, if not already included:**
```
x@x-v:~/src/minikube$ ./out/minikube start --driver=docker --listen-address='0.0.0.0' --apiserver-ips=172.28.24.96 
😄  minikube v1.19.0-beta.0 on Ubuntu 20.04
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🔥  Creating docker container (CPUs=2, Memory=2200MB) ...
💡  minikube is not meant for production use. You are opening non-local traffic
❗  Listening to 0.0.0.0. This is not recommended and can cause a security vulnerability. Use at your own risk
🐳  Preparing Kubernetes v1.20.2 on Docker 20.10.5 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
x@x-v:~/src/minikube$ docker ps
CONTAINER ID        IMAGE                                 COMMAND                  CREATED              STATUS              PORTS                                                                                                                        NAMES
1ff24847a905        gcr.io/k8s-minikube/kicbase:v0.0.19   "/usr/local/bin/entr…"   About a minute ago   Up About a minute   0.0.0.0:32826->22/tcp, 0.0.0.0:32825->2376/tcp, 0.0.0.0:32824->5000/tcp, 0.0.0.0:32823->8443/tcp, 0.0.0.0:32822->32443/tcp   minikube
x@x-v:~/src/minikube$ minikube stop
✋  Stopping node "minikube"  ...
🛑  Powering off "minikube" via SSH ...
🛑  1 nodes stopped.
x@x-v:~/src/minikube$ minikube start
😄  minikube v1.18.1 on Ubuntu 20.04
✨  Using the docker driver based on existing profile
👍  Starting control plane node minikube in cluster minikube
🔄  Restarting existing docker container for "minikube" ...
🐳  Preparing Kubernetes v1.20.2 on Docker 20.10.5 ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v4
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
x@x-v:~/src/minikube$ docker ps
CONTAINER ID        IMAGE                                 COMMAND                  CREATED             STATUS              PORTS                                                                                                                        NAMES
1ff24847a905        gcr.io/k8s-minikube/kicbase:v0.0.19   "/usr/local/bin/entr…"   3 minutes ago       Up About a minute   0.0.0.0:32831->22/tcp, 0.0.0.0:32830->2376/tcp, 0.0.0.0:32829->5000/tcp, 0.0.0.0:32828->8443/tcp, 0.0.0.0:32827->32443/tcp   minikube
```

### PR DESCRIPTION
**Pull request by zhan9san**
Hi,
This change will have apiserver exposed on a fixed port, even though minikube is restarted.
Would it be possible to review and approve this change?

Fixes #11041 
Thanks

### CODE REVIEW

**Comment by zhan9san** [comment]
Hi,
I signed it.
The email for CNCF is the same as what I use in Github Commit message.
Could you help check CLA issue?

**Comment by medyagh** [comment]
@zhan9san how would this work with multiple clusters ? wouldn't they conflict if they have two clusters on same machine?

File: pkg/drivers/kic/oci/oci.go [hunk 1]
```diff
@@ -502,8 +502,13 @@ func generatePortMappings(portMappings ...PortMapping) []string {
 	result := make([]string, 0, len(portMappings))
 	for _, pm := range portMappings {
 		// let docker pick a host port by leaving it as ::
-		// example --publish=127.0.0.17::8443 will get a random host port for 8443
-		publish := fmt.Sprintf("--publish=%s::%d", pm.ListenAddress, pm.ContainerPort)
+		// example --publish=127.0.0.17:8443:8443 will get a fixed host port for 8443
+		publish := ""
+		if pm.HostPort != 0 {
```
Comment (medyagh) [inline pkg/drivers/kic/oci/oci.go:507]: how about add a Check here to verify this port is Free and Usable and if it is not Free and usable return an error to the user.

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/drivers/kic/oci/oci.go (hunk 1 above)
Comment (zhan9san) [reply pkg/drivers/kic/oci/oci.go:507]: It's OK.
I'll add this validation based on what docker does.

**Comment by medyagh** [comment]
@zhan9san are you still working on this ?

**Comment by zhan9san** [comment]
Hi @medyagh 
Thanks for your time.
The validation is added in outer layer instead of driver layer.
I find the implementation of PortAllocator in [moby](https://github.com/moby/moby/blob/master/vendor/github.com/docker/libnetwork/portallocator/portallocator.go) is too complex.

For simplicity, I choose net.Listen to verify whether the port is free and usable.

Please let me know what you think.

File: cmd/minikube/cmd/start.go [hunk 2]
```diff
@@ -1253,6 +1253,15 @@ func validateListenAddress(listenAddr string) {
 func validateListenAPIServerPort(listenAPIServerPort int) {
 	if listenAPIServerPort < 0 || listenAPIServerPort > 65535 {
 		exit.Message(reason.Usage, "Sorry, the port provided with the --listen-apiserver-port flag is invalid: {{.listenAPIServerPort}}.", out.V{"listenAPIServerPort": listenAPIServerPort})
+	} else {
+		ln, err := net.Listen("tcp", fmt.Sprintf(":%d", listenAPIServerPort))
+		if err != nil {
+			exit.Message(reason.Usage, "Sorry, the port provided with the --listen-apiserver-port flag is already allocated: {{.listenAPIServerPort}}.", out.V{"listenAPIServerPort": listenAPIServerPort})
+		}
+		err = ln.Close()
```
Comment (medyagh) [inline cmd/minikube/cmd/start.go:1261]: Doesn’t this need to be defer ?

**Review by medyagh — CHANGES_REQUESTED** [review]

File: cmd/minikube/cmd/start.go (hunk 2 above)
Comment (zhan9san) [reply cmd/minikube/cmd/start.go]: Thanks for pointing it out.

The code snippet above is replaced by `defer`.

By introducing defer statements we can ensure that the socket is always closed.

**Comment by zhan9san** [comment]
Hi @medyagh 
Please hold on.

I'd like to create another PR about FAQ document instead of adding redundant logic in source code.

#11193 and #9404 bring me back to considering whether #11070 is necessary.

Please let me know what you think.

Here is the test of `--ports=8443:8443`

1. On remote Ubuntu server, create a k8s cluster 
```
x@x-v:~$ minikube start --listen-address=0.0.0.0 --apiserver-ips=172.28.24.96 --ports=8443:8443
😄  minikube v1.19.0 on Ubuntu 20.04
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
    > gcr.io/k8s-minikube/kicbase...: 357.67 MiB / 357.67 MiB  100.00% 2.30 MiB
🔥  Creating docker container (CPUs=2, Memory=2200MB) ...
💡  minikube is not meant for production use. You are opening non-local traffic
❗  Listening to 0.0.0.0. This is not recommended and can cause a security vulnerability. Use at your own risk
🐳  Preparing Kubernetes v1.20.2 on Docker 20.10.5 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
```
2. On localhost, copy `~/.kube/config`, `certificate-authority`, `client-certificate` and `client-key` from remote server to corresponding local directory
3. On localhost, change the apiserver from default value(https://192.168.49.2:8443) to specific one(https://172.28.24.96:8443).
4. On localhost, change the location of `certificate-authority`, `client-certificate` and `client-key` in `~/.kube/config` to corresponding local paths.
5. On localhost, run test
```
~/Source/k8s-test/svc ❯  kubectl apply -f test.yml
deployment.apps/my-deployment created
service/my-np-service created
~/Source/k8s-test/svc 5m 50s ❯ kubectl get po                                                                                                        13:15:11
NAME                             READY   STATUS    RESTARTS   AGE
my-deployment-6ccc959b54-nb8p2   1/1     Running   0          11m
~/Source/k8s-test/svc ❯ kubectl get svc                                                                                                              13:16:23
NAME            TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
kubernetes      ClusterIP   10.96.0.1      <none>        443/TCP          16m
my-np-service   NodePort    10.96.152.82   <none>        8080:30080/TCP   11m
```
test.yml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deployment
spec:
  selector:
    matchLabels:
      app: metrics
      department: engineering
  replicas: 1
  template:
    metadata:
      labels:
        app: metrics
        department: engineering
    spec:
      containers:
      - name: hello
        image: "gcr.io/google-samples/hello-app:2.0"
        env:
        - name: "PORT"
          value: "8080"

---
apiVersion: v1
kind: Service
metadata:
  name: my-np-service
spec:
  type: NodePort
  selector:
    app: metrics
    department: engineering
  ports:
  - protocol: TCP
    port: 8080
    nodePort: 30080

```

File: cmd/minikube/cmd/start.go [hunk 3]
```diff
@@ -1244,6 +1248,20 @@ func validateListenAddress(listenAddr string) {
...
+// is valid
+func validateListenAPIServerPort(listenAPIServerPort int) {
+	if listenAPIServerPort < 0 || listenAPIServerPort > 65535 {
+		exit.Message(reason.Usage, "Sorry, the port provided with the --listen-apiserver-port flag is invalid: {{.listenAPIServerPort}}.", out.V{"listenAPIServerPort": listenAPIServerPort})
+	} else {
+		ln, err := net.Listen("tcp", fmt.Sprintf(":%d", listenAPIServerPort))
+		if err != nil {
+			exit.Message(reason.Usage, "Sorry, the port provided with the --listen-apiserver-port flag is already allocated: {{.listenAPIServerPort}}.", out.V{"listenAPIServerPort": listenAPIServerPort})
+		}
+		defer ln.Close()
```
Comment (medyagh) [inline cmd/minikube/cmd/start.go:1261]: this defer needs to happen before Exit

**Review by medyagh — CHANGES_REQUESTED** [review]

File: cmd/minikube/cmd/start.go (hunk 3 above)
Comment (zhan9san) [reply cmd/minikube/cmd/start.go:1261]: Hi @medyagh 
Thanks so much for your time.

If it goes into `exit.Message`, that means the socket wouldn't be created. Do we still need to modify this snippet?
You're far more familiar to Golang that me. I referred the logic in [this file, example_test.go](https://golang.org/src/net/example_test.go), at line 23.

Besides, I noticed #9404 has achieved this feature. The only drawback is that #9404 exposes one more random port.
Should we still work on this feature, `--listen-apiserver-port`?

Please let me know what you think.

**Comment by zhan9san** [comment]
It seems there are some issues in CI workflow.

File: cmd/minikube/cmd/start.go (hunk 3 above)
Comment (spowelljr) [reply cmd/minikube/cmd/start.go:1261]: @zhan9san You are correct about `defer`, if an error is returned no action is required on the socket.

**Comment by marwatk** [comment]
Is anyone still working on this? If not I can try to drive it forward.

We configure our developer machines via ansible and the dynamic k8s port means we have to do a lookup instead of using a fixed value. Not the end of the world, but would be much easier to use a fixed port.

**Comment by zhan9san** [comment]
Hi @marwatk 

Could you please verify does this workaround satisfy your requirement?
https://github.com/kubernetes/minikube/pull/11070#issuecomment-826511944

**Comment by marwatk** [comment]
@zhan9san, thanks for the suggestion. I tried that first and while it works for communication it still puts the random port in `~/.kube/config`. Our (current) workflow needs them to match as our ansible installs and configures minikube (where the lookup is currently needed) and post-install tooling relies on `~/.kube/config` as a source of truth.

**Comment by zhan9san** [comment]
Hi @marwatk 

Could you please provide the minikube version and kubeconfig?

I tried again(v1.22.0), it does show the fixed port, but internal IP address in apiserver.

As `client-certificate`, `client-key` and `certificate-authority` need substituting, you can use Ansible template or `yq` to generate an available kubeconfig file.

**Comment by marwatk** [comment]
@zhan9san, I'm on 1.22:

```bash
$ minikube start \
>     --driver=docker \
>     --embed-certs \
>     --ports=80:80 \
>     --ports=443:443 \
>     --ports=42376:2376 \
>     --ports=48443:8443
😄  minikube v1.22.0 on Ubuntu 18.04
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=12800MB) ...
❗  This container is having trouble accessing https://k8s.gcr.io
💡  To pull new external images, you may need to configure a proxy: https://minikube.sigs.k8s.io/docs/reference/networking/proxy/
🐳  Preparing Kubernetes v1.21.2 on Docker 20.10.7 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ cat ~/.kube/config
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <SNIP>
    extensions:
    - extension:
        last-update: Thu, 09 Sep 2021 12:34:39 MDT
        provider: minikube.sigs.k8s.io
        version: v1.22.0
      name: cluster_info
    server: https://127.0.0.1:49154 <---note the 49154 instead of 48443 here
  name: minikube
<SNIP>
```

**Comment by zhan9san** [comment]
Hi @marwatk 

I run the exactly the same command with you, but still cannot reproduce your issue.

Could you please rename/move current kubeconfig and re-run `minikube start xxx`?


Once it is done, run `docker port minikube`, and show me the output
```bash
$ docker port minikube
8443/tcp -> 0.0.0.0:48443
8443/tcp -> :::48443
8443/tcp -> 127.0.0.1:49313
22/tcp -> 127.0.0.1:49317
2376/tcp -> 127.0.0.1:49316
2376/tcp -> 0.0.0.0:42376
2376/tcp -> :::42376
32443/tcp -> 127.0.0.1:49314
443/tcp -> 0.0.0.0:443
443/tcp -> :::443
5000/tcp -> 127.0.0.1:49315
80/tcp -> 0.0.0.0:80
80/tcp -> :::80
```

**Comment by marwatk** [comment]
Sure enough, deleting my `~./kube/config` prior to starting `minikube` results in the proper port. Super weird because `minikube delete` removes that cluster from the config. And now I can't duplicate it even with an existing. Wondering if it has to do with golang randomly sorting maps and it just picking the first one it gets, but I ran a few in a loop and never saw the behavior I was seeing. :shrug:

At any rate, we're moving to `--driver=native` for our use case, so this is going to be a non-issue for us.

Thanks for your followup!

**Pull request closed without merging**